flink-runtime/src/test/java/org/apache/flink/runtime/security/FlinkSecurityManagerTest.java

```diff
@@ -0,0 +1,415 @@
+        flinkSecurityManager.monitorUserSystemExit();
+        flinkSecurityManager.checkExit(TEST_EXIT_CODE);
+    }
+
+    @Test
+    public void testToggleUserExit() {
+        flinkSecurityManager =
+                new FlinkSecurityManager(ClusterOptions.UserSystemExitMode.THROW, null);
+        flinkSecurityManager.checkExit(TEST_EXIT_CODE);
+        flinkSecurityManager.monitorUserSystemExit();
+        try {
+            flinkSecurityManager.checkExit(TEST_EXIT_CODE);
+            fail();
+        } catch (UserSystemExitException ignored) {
+        }
+        flinkSecurityManager.unmonitorUserSystemExit();
+        flinkSecurityManager.checkExit(TEST_EXIT_CODE);
+    }
+
+    @Test
+    public void testPerThreadThrowUserExit() throws Exception {
+        flinkSecurityManager =
+                new FlinkSecurityManager(ClusterOptions.UserSystemExitMode.THROW, null);
+        ExecutorService executorService = Executors.newSingleThreadExecutor();
+        // Async thread test before enabling monitoring ensures it does not throw while prestarting
+        // worker thread, which is to be unmonitored and tested after enabling monitoring enabled.
+        CompletableFuture<Void> future =
+                CompletableFuture.runAsync(
+                        () -> flinkSecurityManager.checkExit(TEST_EXIT_CODE), executorService);
+        future.get();
+        flinkSecurityManager.monitorUserSystemExit();
+        try {
+            flinkSecurityManager.checkExit(TEST_EXIT_CODE);
+            fail();
+        } catch (UserSystemExitException ignored) {
+        }
+        // This threaded exit should be allowed as thread is not spawned while monitor is enabled.
+        future =
+                CompletableFuture.runAsync(
+                        () -> flinkSecurityManager.checkExit(TEST_EXIT_CODE), executorService);
+        future.get();
+    }
+
+    @Test
+    public void testInheritedThrowUserExit() throws Exception {
+        flinkSecurityManager =
+                new FlinkSecurityManager(ClusterOptions.UserSystemExitMode.THROW, null);
+        flinkSecurityManager.monitorUserSystemExit();
+        try {
+            flinkSecurityManager.checkExit(TEST_EXIT_CODE);
+            fail();
+        } catch (UserSystemExitException ignored) {
+        }
+        CheckedThread thread =
+                new CheckedThread() {
+                    @Override
+                    public void go() {
+                        try {
+                            flinkSecurityManager.checkExit(TEST_EXIT_CODE);
+                            fail();
+                        } catch (UserSystemExitException ignored) {
+                        } catch (Throwable t) {
+                            fail();
+                        }
+                    }
+                };
+        thread.start();
+        thread.sync();
+    }
+
+    @Test
+    public void testLogUserExit() {
```

Comment:
`testLogConfiguration` covers the same scenario - we could delete that test in my opinion.

Reply:
You can see some difference. `test*Configuration` is a little more end-to-end via configuration setting (testing `fromConfiguration`), while others are not using `Configuration` rather directly using constructor.